posts/change-data-capture-cloudsql-postgres-with-apache-airflow-1.md

```diff
@@ -0,0 +1,1002 @@
+
+We'll outline all steps in the process from end to end, from setup to extract. In totality, this includes topics of:
+
+- Setting up security and IAM in GCP
+- Calling the CloudSQL Export API
+- Setting up Airflow and connecting it to GCP
+- Building a DAG in Airflow
+- Configuring Watermarking for data extract intervals
+- Writing an operator in Airflow to extract the data
+- Integrating the operator into the DAG
+- Scheduling and coordinating the extract
+- Deployment to Production on Astronomer Cloud
+
+In a later post, I will show you how to transform or clean the data then add it to a warehouse as part of the same workflow. We will be using the [Astronomer Platform](https://www.astronomer.io/), as it is the easiest way to get your Apache Airflow instance up and running.
+
+## How it Works
+
+The base method for extract uses the [CloudSQL Export API](https://cloud.google.com/sql/docs/postgres/import-export/exporting#rest-v1beta4). This runs an operation with a SQL query, sending the results to a Google Cloud Storage (GCS) bucket as declared by you. You could just as easily extract directly from your database if your system and workload allows for such. 
+
+Using the extract API provides a robust and easily configurable security layer controlled by a service account, such that no system outside of CloudSQL is touching your production database. It also provides an option for a "serverless offload" option in the case that you need to guarantee zero load on your database from the extract. 
+
+The caveat for the export API in general is you [cannot run concurrent CloudSQL Operations](https://cloud.google.com/sql/docs/postgres/import-export/exporting#too-long-rca), so you want to run each query in serial, a scenario perfectly handled by using pools in Airflow.
+
+In general, the API call looks like this:
+
+```bash
+POST https://www.googleapis.com/sql/v1beta4/projects/<project-id>/instances/<instance-id>/export
+{
+ "exportContext":
+   {
+      "fileType": "CSV",
+      "uri": "gs://bucket_name/path_to_csv_file",
+      "databases": ["database_name"],
+      "offload": false
+      "csvExportOptions":
+       {
+           "selectQuery":"SELECT Query - no CTE allowed - subquery okay"
+       }
+   }
+}
+```
+
+This will query `database_name` (only one allowed at a time) using the enclosed `selectQuery` and format the results to a CSV at the GCS bucket and keypath set in `uri`. 
+
+  > *Note: This CSV will not have headers, and in turn, you will also run a second operation to obtain the table schema, which will be described below*
+
+## Step 1: Set Up Airflow
+
+The [Astronomer CLI](https://github.com/astronomer/astro-cli) is the easiest way to run Apache Airflow on your machine. From the CLI, you can establish a local testing environment regardless of where you'll be deploying to from there.
+
+There are two ways to install any version of the Astronomer CLI:
```

Comment:
Think we can cut out a lot of this and just link out to the CLI install doc- feels like all of this content distracts from the core messaging of the post and can be offloaded. I'd remove this line down and just add:

To install the CLI, follow our [install guides](link).

Reply:
Delete the whole section? I added it in the teaching that people clicking away from the page can be bad. What about, "If you already have the Astronomer CLI installed, you can skip to Step ___"

Reply:
I think we can do something in between. For instance, we could just include the Homebrew command for the latest version and say "For alternative installation methods, read our install guide." We then jump straight to the command that confirms the install was successful.

Reply:
Yeah, this ^ works. Just want to get to the good stuff as quickly as possible.

Reply:
Went for a balance between @jwitz and @petedejoy here

![Screen Shot 2021-01-21 at 1 13 39 PM](https://user-images.githubusercontent.com/307956/105413253-b27ddd00-5bea-11eb-9b30-81a1c0c8565e.png)